src/submission/exceptions/exceptions.ts

```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright (c) 2023 The Ontario Institute for Cancer Research. All rights reserved
+ *
+ * This program and the accompanying materials are made available under the terms of
+ * the GNU Affero General Public License v3.0. You should have received a copy of the
+ * GNU Affero General Public License along with this program.
+ *  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT
+ * SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+ * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
+ * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+import { entities as dictionaryEntities } from '@overturebio-stack/lectern-client';
+import { DataRecord } from '@overturebio-stack/lectern-client/lib/schema-entities';
+import _ from 'lodash';
+import { ClinicalEntitySchemaNames } from '../../common-model/entities';
+import entityExceptionRepository from '../../exception/repo/entity';
+import programExceptionRepository from '../../exception/repo/program';
+import { RepoError } from '../../exception/repo/types';
+import { EntityException, ExceptionRecord, ProgramException } from '../../exception/types';
+import { isEntityException, isProgramException } from '../../exception/util';
+
+/**
+ * Check if a valid exception exists and the record value matches it.
+ * If there's a match, we allow the value to pass schema validation.
+ * Filtered schema validation errors and the normalized record value are returned.
+ *
+ * Normalizing is setting the value to start Upper case and to trim whitespace.
+ *
+ * @param programId
+ * @param record
+ * @param schemaValidationErrors
+ */
+export const checkForProgramAndEntityExceptions = async ({
+  programId,
+  record,
+  schemaName,
+  schemaValidationErrors,
+}: {
+  programId: string;
+  record: DataRecord;
+  schemaName: ClinicalEntitySchemaNames;
+  schemaValidationErrors: dictionaryEntities.SchemaValidationError[];
+}) => {
+  const filteredErrors: dictionaryEntities.SchemaValidationError[] = [];
+  let normalizedRecord = record;
+
+  // retrieve submitted exceptions for program id (both program level and entity level)
+  const { programException, entityException } = await queryForExceptions(programId);
```

Comment:
Though this is not strictly speaking "wrong", I'm urged to suggest not using functions/variables before they're defined. I had to do `search` to find this in line 120, so I could find whether there's a try/catch associated with this await.